Debug: create private code memories per store when debugging is enabled.

cfallin opened the pull request:
This will allow patching code to implement e.g. breakpoints. (That is, for now the copies are redundant, but soon they will not be.)

This change follows the discussion [here] and offline to define a few types that better encapsulate the distinction we want to enforce. Basically, there is almost never a bare `CodeMemory`; they are always wrapped in an `EngineCode` or `StoreCode`, the latter being a per-store instance of the former. Accessors are moved to the relevant place so that, for example, one cannot get a pointer to a Wasm function's body without being in the context of a `Store` where the containing module has been registered. The registry then returns a `ModuleWithCode` that boxes up a `Module` reference and `StoreCode` together for cases where we need both the metadata from the module and the raw code to derive something.

The only case where we return raw code pointers to the `EngineCode` directly have to do with Wasm-to-array trampolines: in some cases, e.g. `InstancePre` pre-creating data structures with references to host functions, it breaks our expected performance characteristics to make the function pointers store-specific. This is fine as long as the Wasm-to-array trampolines never bake in direct calls to Wasm functions; the strong invariant is that Wasm functions never execute from `EngineCode` directly. Some parts of the component runtime would also have to be substantially refactored if we wanted to do away with this exception.

The per-`Store` module registry is substantially refactored in this PR. I got rid of the modules-without-code distinction (the case where a module only has trampolines and no defined functions still works fine), and organized the BTreeMaps to key on start address rather than end address, which I find a little more intuitive (one then queries with the dual to the range -- 0-up-to-PC and last entry found).

[here]: https://github.com/bytecodealliance/wasmtime/pull/12051#pullrequestreview-3493711812

alexcrichton commented on crates/wasmtime/src/runtime/module.rs:
```diff
@@ -1133,14 +1143,18 @@ impl Module {
     #[cfg(any(feature = "gc", feature = "debug"))]
-    pub(crate) fn text_offset(&self, pc: usize) -> u32 {
-        u32::try_from(pc - self.inner.module.text().as_ptr() as usize).unwrap()
+    pub(crate) fn text_offset(&self, pc: usize, code_memory: &CodeMemory) -> u32 {
```
Since this doesn't actually use `&self` at all here any more, maybe this could move to `CodeMemory` directly?

cfallin replied:
(Updated after refactor)

cfallin commented:
> Then we can get a "module with default code" from the Module itself if we know we aren't going to return pointers to it.

And, actually, as a refinement on that: we could do some type-system shenanigans so that a "module with default code" is not allowed to return raw code pointers, only other (meta)data from the image, while a "module with code" obtained from an instance is.

fitzgen commented on crates/wasmtime/src/runtime/module/registry.rs:
```diff
@@ -99,42 +105,74 @@ impl ModuleRegistry {
+                    memory.publish()?;
+                    let object = code_object.with_private_code(Arc::new(memory));
+                    v.insert(Arc::new(object))
+                }
```
I originally expected this to be an assert that the entry was not occupied, but I double checked and we do in fact potentially register the same code multiple times (e.g. we register a module's code on each `Instance::new` call and you can obviously instantiate the same module multiple times in the same store).

This subtlety pushes me towards asking for some kind of assertion that the existing private code object matches our `code_object` parameter, but I don't see how that is really possible without a full byte-for-byte comparison of the whole mmap, which seems like it is probably too expensive.

So ultimately maybe we just expand this method's doc comment to note that the method is idempotent and that modules will be registered multiple times in certain scenarios? That reminder/context probably would have let me avoid going down this rabbit hole.

Or maybe we should invert some logic here, and *always* use `self.private_code` (after renaming it to something like `self.registered_code`) so that we can super easily tell if a code object is already registered and return early if so, without needing to actually do the careful text range queries below? I think this inversion would make the idempotency logic much clearer. Thoughts?

cfallin replied:
Yes, the logic here is pre-existing and a little weird, but appears to be necessary because of the way that modules within a component share one compiled artifact (or at least, without a deeper refactor of the whole components runtime). I've added more doc-comments in the refactor on the `ModuleRegistry` that hopefully address this (most of this file is new now as well).

fitzgen reviewed (commented):
To briefly repeat some comments from DM: I am in favor of the refactor discussed in this PR's main thread where

* we create a type separation between the shared code at the engine level and the (potentially private copy) code used at the store level
* all accessors for function pointers and such move to the store-level code type (or a `Module` and store-level code pair type)
* the act of creating the store-level code is the thing one place we are making a private copy or reusing the shared copy

bikeshedding names for these two new types:

* `EngineCode` and `StoreCode`
* `CodeModule` and `CodeInstance`
* `CodeFactory` and `CodeInstance` lol

cfallin commented:
Just an update on progress here: I'm [midway through the refactor](https://github.com/cfallin/wasmtime/tree/code-segments-for-everyone), and it's proving quite a lot more invasive than I had hoped. On the positive side, I've worked out a way to have truly unique ownership of a `StoreCode`, so mutation can be proved to be safe given a mut borrow of the `Store`. But in terms of impact, this is very deeply invasive, and I worry that (for example) indirect function calls are going to be penalized: lazily setting up the funcref now requires looking up the `StoreCode` for the module in question, so rather than a few fast array accesses in the `CompiledModule`, we have a BTree lookup. The impact on the guest profiler is pretty deep (it handles all its metadata regarding PCs at the module level) and `InstancePre` (though the latter may be OK only if we allow executing Wasm-to-array trampolines in the *engine* code rather than the *store* code, allowing pre-resolution of host funcrefs; this breaks a desirable invariant that we never execute engine code directly though).

I'll keep plugging away at this, but just wanted to surface the cost of the above requested refactor :-) I will say that with the types as now defined, I feel much more reasonable about having not missed any place that would execute the wrong copy of code.

cfallin commented on crates/wasmtime/src/runtime/instantiate.rs:
```diff
@@ -150,22 +149,30 @@ impl CompiledModule {
     /// These trampolines are used for array callers (e.g. `Func::new`)
     /// calling Wasm callees.
-    pub fn array_to_wasm_trampoline(&self, def_func_index: DefinedFuncIndex) -> Option<&[u8]> {
+    pub fn array_to_wasm_trampoline<'a>(
```
(Updated after refactor)

cfallin commented on crates/wasmtime/src/runtime/instance.rs:
```diff
@@ -308,7 +309,8 @@ impl Instance {
         // properly referenced while in use by the store.
-        let module_id = store.modules_mut().register_module(module);
+        let engine = store.engine().clone();
+        let (module_id, code_memory) = store.modules_mut().register_module(module, &engine)?;
```
This was borrowing-related but I avoided it in the refactor by putting the `register_*` methods on the `Store` directly so it can pass in the `&Engine` from its `&Arc` separately borrowed -- updated in refactor.

cfallin commented on crates/wasmtime/src/runtime/instance.rs:
```diff
@@ -231,7 +231,8 @@ impl Instance {
         // of funcs-with-holes should generally be empty. As a result the
         // process of filling this out is not super optimized at this point.
-        store.modules_mut().register_module(module);
+        let engine = store.engine().clone();
```
(Updated inrefactor)

cfallin commented on crates/wasmtime/src/runtime/trap.rs:
```diff
@@ -440,7 +440,10 @@ impl FrameInfo {
+                .default_code_memory()
+                .address_map_data(),
             text_offset,
         );
```
(Updated in refactor)

cfallin commented on crates/wasmtime/src/runtime/vm/mmap_vec.rs:
```diff
@@ -270,6 +270,58 @@ impl MmapVec {
+        // position-in-the-4K-page of the reference and referred-to
+        // address. This is true today because we only generate
+        // "colocated" calls to other functions in the module, and we
+        // never use the `func_addr` or `symbol_addr` opcodes.
```
Yep, I think we'll want to think carefully about this in its own right -- if you don't mind, could we defer alignment-related things (since it appears to be a pre-existing issue as well)?

fitzgen replied:
Yeah for sure, was not suggesting anything that needs to be dealt with here and now.

cfallin commented:
OK, I've now refactored in the direction we discussed above -- I feel a lot better about the types ensuring we're not leaking the wrong kind of function pointers (i.e. not executing the private copy) now!

I've updated the initial comment based on the new comment message above -- please see for some of the caveats and reasoning. Let me know what you think!

alexcrichton commented on crates/wasmtime/src/runtime/code.rs:
```diff
@@ -100,3 +284,180 @@ impl From<Arc<ComponentTypes>> for Types {
+        // of breakpoints (by code patching) independently in each
+        // one.
+        let code = if engine.tunables().debug_guest {
+            let mut private_copy = engine_code.original_code.deep_clone(engine)?;
```
As a follow-up and/or pending issue, it might be nice to only clone the text section here as opposed to anything else since everything else in the module is irrelevant for the cloning purpose. Modules can often be 2x larger than the text section for the `.wasmtime.addrmap` section alone so we might have some nice wins cutting down on the size. My hope is that this'd be pretty easy refactor to tack on in the future though where the only caveat would be to redirect some accesses of metadata to the `EngineCode` layer, if any, rather than using it through `StoreCode` (and this may already very well be solved entirely in the refactorings of this PR)

alexcrichton reviewed (commented):
To me this all looks great, thanks for pushing on the refactor!

> this is very deeply invasive, and I worry that (for example) indirect function calls are going to be penalized

I didn't run across this myself, so I wanted to clarify -- is this still present or did this end up getting resolved during refactoring?

cfallin commented on crates/wasmtime/src/runtime/code.rs:
```diff
@@ -100,3 +284,180 @@ impl From<Arc<ComponentTypes>> for Types {
+    ///
+    /// This is the only reference to this CodeMemory. The StoreCode
+    /// is owned directly by the Store's ModuleRegistry.
+    Private(Box<CodeMemory>),
```
Done!

cfallin commented on crates/wasmtime/src/runtime/instantiate.rs:
```diff
@@ -65,7 +68,9 @@ impl CompiledModule {
         // so we may be registering the same code range multiple times here.
-        profiler.register_module(&self.code_memory.mmap()[..], &|addr| {
+        //
+        // TODO: register for each StoreCode.
```
Done! Filed #12105.

cfallin replied:
Hmm, actually, this seems to cause a test failure: the registration always happens, so we can't simply throw an error. I think maybe the best we can do (short of a refactor that carries the `ProfilingAgent` through to the point of instantiation, not just module loading, and registers each instantiation) is have the caveat that `perf record` will not have registered profiling info for debugging-private copies of code. This may be fine (given issue to track)?

cfallin commented on crates/wasmtime/src/runtime/module/registry.rs:
```diff
@@ -1,173 +1,198 @@
+    /// Every module is placed in this map, but not every module will
+    /// be in a LoadedCode entry, because the module may have no text.
+    modules: BTreeMap<RegisteredModuleId, Module>,
 }
```
Sure -- looks like about a 3% hit on a small module, sequential instantiation:

```
% cargo bench --bench instantiation -- sequential/pooling/wasi.wasm

[ baseline run ]

sequential/pooling/wasi.wasm
                        time:   [1.4748 µs 1.4770 µs 1.4795 µs]

[ run with this PR's changes ]

sequential/pooling/wasi.wasm
                        time:   [1.5173 µs 1.5184 µs 1.5196 µs]
                        change: [+2.9095% +3.0865% +3.2436%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

Since this is constant-overhead-per-module, the operative number is the extra ~40 ns; that seems OK to me?

cfallin commented:
OK, I think this is ready for another look -- thanks @alexcrichton!

fitzgen commented:
> Actually, let's benchmark -- Sightglass says:

2 processes x 5 iterations per process = 10 total samples per group. You can't generally do significance tests with less than 30 samples per group (and you probably want more than that for the `cycles` measure, since it is so noisy).

If you're going to benchmark, then you should do it properly :-p

If the amount of time taken to run the benchmarks was a concern, then you could pass `--benchmark-phase execution` to avoid re-compiling on each iteration.

Also, fwiw, we have some call-indirect specific micro benchmarks [here](https://github.com/bytecodealliance/wasmtime/blob/196f09fe95efeae48724344b63c8b2182e127ce0/benches/call.rs#L901) which might be better for measuring this specific thing.

cfallin commented:
OK, cool, I'll rerun this in a bit. Indeed I only had a small sliver of time last night and didn't want to wait hours for results. Good to know that one can re-use compilations, thanks.

My intent with using the real-program benchmarks rather than microbenchmarks was to measure the effect of call-indirects in proper context, but I can do the microbenchmarks too.

fitzgen commented:
> didn't want to wait hours for results

If I do a full run of `default.suite` it takes 27 mins on my old thinkpad, probably faster for you on a newer MBP, fwiw. Also faster if doing `--benchmark-phase execution` or if measuring instructions retired and you can drop to a small handful of iterations. But yeah 27 mins is not great.

> So perhaps the ~1% overhead is in the additional plumbing in general on the table funcref init path; that's unavoidable unless we monomorphize the whole funcref-related runtime (libcalls on down) on debug and no-debug versions I think. Maybe we eat the 1%. What do you think?

This is only the case for lazy table init, right? If it isn't lazy then we'd still be hitting these branches, but at instantiation time instead, and we would be doing them all at once, so presumably the branch predictor would be more useful in that case, no?

If this is only for lazy table init, then yeah seems fine by me. Although maybe Alex has more ideas here. I originally was thinking that the module registry could possibly contain a hash map instead of a btree map for the new member, but it sounds like the costs are not related to the data structure.

Slightly more concerning if this is a 1% slowdown regardless of lazy table init...

cfallin commented:
> This is only the case for lazy table init, right? 

It seems so, yeah -- flipping the default to eager table init in tunables, I get (100 data points, 10 proc * 10/proc):

```
execution :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  No difference in performance.

  [453510422 463605372.90 509169192] new-eager-init.so
  [453961879 467730247.66 531244575] old-eager-init.so
```

alexcrichton reviewed (approved):
Code all looks good to me, thanks for the updates!

Questions on perf though, so it sounds like:

* Instantiation-wise this is measured as a 3% regression.
* call_indirect-wise it's measured as a "no meaningful change"

Is that right?

I would view this as adding relatively low-hanging fruit to the "perf tree" to pick in the future, and to me it's mostly a question of whether we pick that fruit here or not. The extra data structures can likely be optimized with fewer lookups/insertions or something like that when debugging enabled is my thinking.

We're also not really measuring heap impact (IIRC B-Trees are relatively bad for that?) which might be another consideration here. Basically I would be surprised if these exact data structures lived as-is for a long period of time past perf testing and such. I'd be more comfortable if the addition of debugging-related features didn't impact the data structures when debugging wasn't turned on, but it's a pretty minor impact here.

Basically I'm ok with this personally, but I think if we land this as-is it's worth opening an issue describing the possible optimization opportunity. More-or-less it feels like we shouldn't need 4 btree insertions when a single module is instantiated in a store in the fast path. That feels like it's pretty overkill relative to "lightweight instantiation". I realize we probably already have 3 insertions or something like that today, but this is basically a ripe area for optimization in the future if someone's interested.

cfallin commented:
Followup filed at #12111.
